### PR TITLE
Update dependencies

### PR DESCRIPTION
**Pull request by mvz**
- Update to RubuCop 0.37.0
- Fix new offenses
- Sort RuboCop configuration
- Update Aruba dependency

Just good housekeeping :smile:.

### CODE REVIEW

**Comment by mvz** [comment]
As part of reviewing this, please consider the items marked for discussion in the rubocop configuration.

File: .rubocop.yml [hunk 1]
```diff
@@ -4,6 +4,10 @@ AllCops:
     - 'tmp/**/*'
     - 'vendor/**/*'
 
+# FIXME: Re-enable once private_attr_reader is no longer used
+Lint/UselessAccessModifier:
```
Comment (troessner) [inline .rubocop.yml]: So in the latest version

```Ruby
private_attr_reader :foo
````

is marked with that?

File: .rubocop.yml [hunk 2]
```diff
@@ -22,21 +26,19 @@ Metrics/LineLength:
...
-# Allow small arrays of words with quotes
-Style/WordArray:
-  MinSize: 3
+Style/AccessorMethodName:
+  Exclude:
+    - 'lib/reek/context/visibility_tracker.rb'
 
-# Allow single-line method definitions
-Style/SingleLineMethods:
+# TODO: Discuss
```
Comment (troessner) [inline .rubocop.yml]: :+1:

File: .rubocop.yml [hunk 3]
```diff
@@ -46,23 +48,42 @@ Style/DotPosition:
 Style/EmptyLineBetweenDefs:
   AllowAdjacentOneLineDefs: true
 
+# Maintain compatibility with older Rubies
```
Comment (troessner) [inline .rubocop.yml]: But the freeze pragma will just be ignored by older rubies, won't it?
I thought about adding the pragma everywhere as well ...

File: .rubocop.yml [hunk 4]
```diff
@@ -46,23 +48,42 @@ Style/DotPosition:
...
+# Indent one level for follow-up lines
+Style/MultilineMethodCallIndentation:
+  EnforcedStyle: indented
+
+# Indent one level for follow-up lines
+Style/MultilineOperationIndentation:
+  EnforcedStyle: indented
+
+# TODO: Discuss
+Style/MutableConstant:
```
Comment (troessner) [inline .rubocop.yml]: :+1:

File: .rubocop.yml [hunk 5]
```diff
@@ -46,23 +48,42 @@ Style/DotPosition:
...
+  Enabled: false
+
+# TODO: Discuss
+Style/ParallelAssignment:
+  Enabled: false
```
Comment (troessner) [inline .rubocop.yml]: :+1: as well. I always hated this rule.

File: .rubocop.yml [hunk 6]
```diff
@@ -46,23 +48,42 @@ Style/DotPosition:
...
-    - 'lib/reek/ast/sexp_extensions/send.rb'
-    - 'lib/reek/ast/sexp_extensions/super.rb'
-    - 'lib/reek/ast/sexp_extensions/variables.rb'
-
-Style/AccessorMethodName:
-  Exclude:
-    - 'lib/reek/context/visibility_tracker.rb'
+# Always use raise to signal exceptions
+Style/SignalException:
+  EnforcedStyle: only_raise
```
Comment (troessner) [inline .rubocop.yml]: :-1: I like the differentiation between `fail` and `raise`

File: .rubocop.yml [hunk 7]
```diff
@@ -46,23 +48,42 @@ Style/DotPosition:
...
 
-Style/ParallelAssignment:
+# Allow single-line method definitions
+Style/SingleLineMethods:
   Enabled: false
```
Comment (troessner) [inline .rubocop.yml]: :+1:

File: lib/reek/ast/node.rb [hunk 8]
```diff
@@ -109,14 +109,14 @@ def look_for_type(target_type, ignoring = [], &blk)
         each_sexp do |elem|
           elem.look_for_type(target_type, ignoring, &blk) unless ignoring.include?(elem.type)
         end
-        blk.call(self) if type == target_type
+        yield self if type == target_type
```
Comment (troessner) [inline lib/reek/ast/node.rb]: Yep, that's nicer!

**Comment by troessner** [comment]
Reviewed, looks good, couple of discussion items around rubocop and then lgtm

File: .rubocop.yml (hunk 1 above)
Comment (mvz) [inline .rubocop.yml]: No, we have `private` before each of them and that triggers this cop.

File: .rubocop.yml (hunk 3 above)
Comment (mvz) [inline .rubocop.yml]: I tried this on another project in the mean time, and I agree. I'll change the enforced style to `always`.

File: .rubocop.yml (hunk 2 above)
Comment (mvz) [inline .rubocop.yml]: Another option is to settgis to RuboCop's old default.

File: .rubocop.yml (hunk 6 above)
Comment (mvz) [inline .rubocop.yml]: This setting was only moved to sort alphabetically.

File: .rubocop.yml (hunk 5 above)
Comment (mvz) [inline .rubocop.yml]: Ah, this also was just moved, but I added the discuss bit to hopefully get some explanatory comment for it. I suppose I can just add `# @troesnerr always hated this rule` :laughing:.

Comment (troessner) [inline .rubocop.yml]: :laughing:

**Comment by mvz** [comment]
Updated adding string freeze comments

File: Rakefile [hunk 9]
```diff
@@ -1,3 +1,4 @@
+# frozen_string_literal: true
```
Comment (troessner) [inline Rakefile]: Can we pls have a newline and here and for all other files below? I really think this would increase readability.
Also, can we make this a separate commit?

Comment (mvz) [inline Rakefile]: > Can we pls have a newline and here and for all other files below? I really think this would increase readability.

These changes are all autogenerated by RuboCop, so I'd rather not :smile:.

> Also, can we make this a separate commit?

Sure.

File: .rubocop.yml (hunk 6 above)
Comment (mvz) [inline .rubocop.yml]: We currently only actually use `raise` in Reek. Is there any place you would like to use `fail`. As a side note, RuboCop actually changed the default to `only_raise` recently.

Comment (troessner) [inline .rubocop.yml]: >> Is there any place you would like to use fail

We haven't written any "fails" / "raises" ever since we introduced rubocop :laughing: 
But yes, grepping for "raise" shows me a couple of places where I would do that. 
On the other hand, that's the same discussion we recently had in [RubyCritic](https://github.com/whitesmith/rubycritic/pull/119) so let me just quote myself (ah, i love quoting myself!!!):

>> Not the biggest fan of this since I really like the semantic difference between "fail" and "raise". 
However, sticking a reasonable style guide project wide is more important than my personal taste here.

So let's just stick with rubocop's defaults ;)

File: .rubocop.yml (hunk 1 above)
Comment (troessner) [inline .rubocop.yml]: But then we could easily fix this by moving those declarations up like [here](https://github.com/troessner/reek/blob/master/lib/reek/source/source_locator.rb#L32) and then remove this configuration setting, can't we?
If yes, I'd prefer to do this in this  PR since it's also only a couple of occurences.

Comment (mvz) [inline .rubocop.yml]: Yes, it is an easy fix. My reasoning was that we will be touching the same locations again in #804, adding back the `private` keyword that would be removed now.

Comment (mvz) [inline .rubocop.yml]: On third thought, I can let RuboCop fix this automatically for now.

File: .rubocop.yml (hunk 2 above)
Comment (mvz) [inline .rubocop.yml]: To expand on this a bit further, RuboCop started out enforcing `alias_method` everywhere, and this is how things stand in Reek at the moment. With the new version it changed to requiring `alias_method` inside methods, and `alias` outside. I have a slight preference for the new style, but I think we pick a style and have RuboCop enforce it either way.

File: .rubocop.yml (hunk 4 above)
Comment (mvz) [inline .rubocop.yml]: So, to be clear, you are against RuboCop enforcing use of `freeze` on literals assigned to constants?

File: .rubocop.yml (hunk 6 above)
Comment (mvz) [inline .rubocop.yml]: Ok, I removed the setting so we use RuboCop's new default (which is `only_raise` so there are no code changes).

File: .rubocop.yml (hunk 1 above)
Comment (mvz) [inline .rubocop.yml]: > I can let RuboCop fix this automatically for now.

Oddly, this smell has no autocorrect :disappointed:.

**Comment by mvz** [comment]
@chastell this is the PR where I'd like your input on the RuboCop settings.

File: .rubocop.yml (hunk 4 above)
Comment (troessner) [inline .rubocop.yml]: No, I'm for it.;)

It's super-hard for me to answer those comments because I just get an email with a link from github. Which doesn't work, cause it refers to an outdated diff. So I have to click on ALL outdated diffs to find your latest comment. @github in the name of all gods available, please fix that.
Let's continue discussing in top level comments.

**Comment by mvz** [comment]
> Let's continue discussing in top level comments.

:+1:

So, to sum up, the following cop settings are/were under discussion:

* Style/MutableConstant: @troessner is for it, and so am I. I wlll enable it.
* Style/Alias: RuboCop's default changed, do we follow it?
* Lint/UselessAccessModifier: Enabled now, code fixed.
* Style/ParallelAssignment: @troessner doesn't like this rule, I'd like to take a look how relevant this is.

**Comment by chastell** [comment]
(Sorry, still firefighting elsewhere.)

My general ideas (I’ve been dealing with this across all of my projects, updating RuboCop in each…):

* my general rule of thumb is to err on the side of RuboCop defaults
  * which is frustrating when they change
  * so it’s tempting to just configure it to pre-change defaults
  * but in the long run, if the change is not something I deeply oppose, I do `rubocop --auto-correct` to the new defaults eventually
* I have a few pet peeves with RuboCop defaults that I consider unacceptable and disable those
  * `Style/AndOr` – `and` + `or` have a different precedence in Ruby than `&&` + `||` _on purpose_ and there’s nothing wrong with using the right ones in the right places
  * `Style/Not` – I find self-standing `not` so much more readable than `!` glued to a predicate
  * `Style/SingleLineBlockParams` – this one is my biggest issue and should be srsly burned with fire, readable block params are so much better than `{ |a, e| … }`
  * `Style/TrailingCommaIn{Arguments,Literal}` – I prefer to enforce the comma in multiline for cleaner diffs
* that said, I’m open for discussion on each and every one (although I do believe there are better uses of our Reek devel time) ;)

As for particulars here (I just realised some of them are not changes but what we have now just moved around…):

* `Style/Alias` – I’d just let RuboCop auto-correct to the current default and get used to it
* `Style/Multiline*CallIndentation` – I assume `EnforcedStyle: indented` makes our code shorter (horizontally), so I’m ok with it :)
* `Style/ParallelAssignment` – I hit a few places in my projects where parallel assignment was the most readable thing to do, so I’m ok with disabling this, but I’m also ok with keeping it on
* `Style/SingleLineMethods` – I don’t like single-line methods, but I think we use them quite often, so I understand we need to disable this :)
* `Style/WordArray` – I love `%w(small word arrays)`, but I understand it takes a bit of eye training to switch to them, so I’m ok with keeping this

**Comment by chastell** [comment]
My brain also says there should be an empty line after

```ruby
# frozen_string_literal: true
```

– but if it’s just me then I’ll cope. ;)

**Comment by mvz** [comment]
> I’m open for discussion on each and every one (although I do believe there are better uses of our Reek devel time) ;)

Agreed, so I'm not going to adjust any of the settings that we already have, and we can defer discussion to when we have had too many beers.

As to the two remaining settings from my list:

* Style/Alias: I agree with @chastell we should just get used to the new default and let RC update our code.
* Style/ParallelAssignment: @troessner doesn't like this rule, @chastell is neutral; I took a look and we 'violate' this cop on one occasion, where I think use of parallel assignment is perfectly acceptable.

@troessner, final thoughts on Style/Alias?

**Comment by troessner** [comment]
>> @troessner, final thoughts on Style/Alias?

Yep, let's go with rubocop defaults ;)

**Comment by mvz** [comment]
> but if it’s just me then I’ll cope. ;)

No, it's not just you. I moved that part to a separate PR: #856.

**Comment by chastell** [comment]
LGTM! @troessner?

(Let’s get this merged so I can drop private_attr without making others rebase their PRs.) :)

**Comment by troessner** [comment]
Merging this bad boy!

**Comment by troessner** [comment]
>> (Let’s get this merged so I can drop private_attr without making others rebase their PRs.) :)

So looking forward to this! One dependency less!
I think in the long run we should only have `parser` as dependency.
